src/Microsoft.AspNetCore.Blazor/RenderTree/RenderTreeDiffBuilder.cs

```diff
@@ -149,19 +149,84 @@ public static void DisposeFrames(RenderBatchBuilder batchBuilder, ArrayRange<Ren
             ref DiffContext diffContext,
             int oldStartIndex, int oldEndIndexExcl,
             int newStartIndex, int newEndIndexExcl)
+        {
+            // The overhead of the dictionary used by AppendAttributeDiffEntriesForRangeSlow is
+            // significant, so we want to try and do a merge-join if possible, but fall back to
+            // a hash-join if not. We'll do a merge join until we hit a case we can't handle and
+            // then fall back to the slow path.
+            //
+            // Also since duplicate attributes are not legal, we don't need to care about loops or
+            // the more complicated scenarios handled by AppendDiffEntriesForRange.
+            //
+            // We also assume that we won't see an attribute occur with different sequence numbers
+            // in the old and new sequences. It will be handled correct, but will generate a suboptimal
+            // diff.
+            var hasMoreOld = oldEndIndexExcl > oldStartIndex;
+            var hasMoreNew = newEndIndexExcl > newStartIndex;
+            var oldTree = diffContext.OldTree;
+            var newTree = diffContext.NewTree;
+
+            while (hasMoreOld || hasMoreNew)
+            {
+                var oldSeq = hasMoreOld ? oldTree[oldStartIndex].Sequence : int.MaxValue;
+                var newSeq = hasMoreNew ? newTree[newStartIndex].Sequence : int.MaxValue;
+                var oldAttributeName = oldTree[oldStartIndex].AttributeName;
+                var newAttributeName = newTree[newStartIndex].AttributeName;
+
+                if (oldSeq == newSeq &&
+                    string.Equals(oldAttributeName, newAttributeName, StringComparison.Ordinal))
+                {
+                    // These two attributes have the same sequence and name. Keep merging.
+                    AppendDiffEntriesForAttributeFrame(ref diffContext, oldStartIndex, newStartIndex);
+
+                    oldStartIndex = NextSiblingIndex(oldTree[oldStartIndex], oldStartIndex);
+                    newStartIndex = NextSiblingIndex(newTree[newStartIndex], newStartIndex);
+                    hasMoreOld = oldEndIndexExcl > oldStartIndex;
+                    hasMoreNew = newEndIndexExcl > newStartIndex;
+                }
+                else if (oldSeq < newSeq)
+                {
+                    // An attribute was removed from the new sequence.
```

Comment:
Typo? I think this should say removed from the *old* sequence

Reply:
Fixed